ml-prims/src/common/buffer_base.hpp

```diff
@@ -38,11 +43,23 @@ class buffer_base
 
     buffer_base& operator=(const buffer_base& other) = delete;
 
-    buffer_base(cudaStream_t stream, size_type n)
-        :  _size(n), _capacity(n), _data(nullptr), _stream(stream)
-    {}
+    buffer_base(std::shared_ptr<Allocator> allocator, cudaStream_t stream, size_type n=0)
+        :  _size(n), _capacity(n), _data(nullptr), _stream(stream), _allocator(allocator)
+    {
+        if ( _capacity > 0 )
+        {
+            _data = static_cast<value_type*>(_allocator->allocate( _capacity*sizeof(value_type), _stream ));
+            CUDA_CHECK( cudaStreamSynchronize( _stream ) );
```

Comment:
Depending on what we decide in #425 we might have to do this inside either the `host_buffer` or keep this here itself.

Reply:
Yes, but I see this as part of the discussion in #425 not this PR. So this is nothing to change in this PR now. Depending on how we conclude #425 a new PR might be needed. Agreed?

Reply:
@teju85 since your reacted with thumbs up on my comment I will resolve this conversation. Can you approve this PR or have I missed a change that you requested?